app/models/concerns/smart_proxy_reference.rb

```diff
@@ -0,0 +1,51 @@
+module SmartProxyReference
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    attr_accessor :smart_proxy_references
+
+    def smart_proxy_reference(hash)
+      raise Foreman::Exception.new(N_("Proxy reference should be { :relation => [:column_name, ...]}")) unless correct_reference_format?(hash)
+      @smart_proxy_references ||= {}
```

Comment:
If you don't mind, I would like to keep just a simple hash because `OpenStruct` is not enumerable I can use `select` or `reject` to find columns on `self` easily.